### PR TITLE
Remove expensive array iteration operations

### PR DESCRIPTION
**Pull request by Ocramius**

### CODE REVIEW

**Comment by Ocramius** [comment]
YOLO :ship:

File: lib/Doctrine/ORM/PersistentCollection.php [hunk 1]
```diff
@@ -240,11 +240,12 @@ public function getSnapshot()
...
     {
-        return array_udiff_assoc(
-            $this->snapshot,
-            $this->collection->toArray(),
-            function($a, $b) { return $a === $b ? 0 : 1; }
-        );
+        $collectionItems = $this->collection->toArray();
+
+        return \array_values(\array_diff_key(
+            \array_combine(\array_map('spl_object_hash', $this->snapshot), $this->snapshot),
```
Comment (drupol) [inline lib/Doctrine/ORM/PersistentCollection.php:246]: Note for later, maybe we should use `spl_object_id` as mentionned here: https://github.com/doctrine/orm/pull/6878

Comment (greg0ire) [reply lib/Doctrine/ORM/PersistentCollection.php:246]: It's in the backport todo list, yes :)

**Comment by carnage** [comment]
Tried pulling this one into latest 2.10.x but it seems that the commits no longer exist in a fresh clone of the repo. Perhaps @Ocramius still has them locally, otherwise I'm stuck to recreating via copy/paste from github as fresh commits.

**Comment by greg0ire** [comment]
I think you won't find them by hash, but you can find them by commit message. At least it was doable last time I tried :)

**Comment by greg0ire** [comment]
Lloks like they were applied several times, in fact :sweat_smile: 
```
git log --oneline --all|grep 'Replacing possible'
c6116d4e5 Replacing possible O(n^2) operation from `PersistentCollection` diffing operations
4fc48a66b Replacing possible O(n^2) operation from `PersistentCollection` diffing operations
200e53a27 Replacing possible O(n^2) operation from `PersistentCollection` diffing operations
```

:thinking: that must be because my clone isn't fresh. Anyway old-prototype-3.x is still up and must have those commits, but has been rebased.

**Comment by simonberger** [comment]
This PR has like 30 changed lines. 8)
What I did in a similar, bigger case is to grab a patch from the PR and apply it.

https://github.com/doctrine/orm/pull/6619.patch

**Comment by carnage** [comment]
I was trying to find an option to grab the patch from the PR yesterday and gave up, thanks for the link :D

**Pull request merged**